### PR TITLE
update samtools/bcftools/htslib to 1.18 with builder stages included

### PR DESCRIPTION
**Pull request by kapsakcj**
~~Setting as a draft until any kinks are worked out.~~ Now ready for review. Dockerfiles restructured and tests are passing.

@erinyoung I kept the tests you added & continued working off of your branch, hope that's ok. Looks like the `plot-ampliconstats` test completed successfully for samtools.

This PR adds dockerfiles for samtools, bcftools, and htslib v1.18. All 3 now have the `builder` stage for code compilation, where previously only samtools 1.17 included a builder stage.

Pull Request (PR) checklist:
- [x] Include a description of what is in this pull request in this message.
- [x] The dockerfile successfully builds to a test target for the user creating the PR. (i.e. `docker build --tag samtools:1.15test --target test docker-builds/samtools/1.15` )
- [x] Directory structure as name of the tool in lower case with special characters removed with a subdirectory of the version number (i.e. `spades/3.12.0/Dockerfile`)
   - [x] (optional) All test files are located in same directory as the Dockerfile (i.e. `shigatyper/2.0.1/test.sh`)
- [x] Create a simple container-specific [README.md](https://github.com/StaPH-B/docker-builds/blob/master/.github/workflow-templates/readme-template.md) in the same directory as the Dockerfile (i.e. `spades/3.12.0/README.md`)
   - [x] If this README is longer than 30 lines, there is an explanation as to why more detail was needed
- [x] Dockerfile includes the recommended [LABELS](https://github.com/StaPH-B/docker-builds/blob/master/dockerfile-template/Dockerfile#L8-L18) 
- [x] Main [README.md](https://github.com/StaPH-B/docker-builds/blob/master/README.md) has been updated to include the tool and/or version of the dockerfile(s) in this PR
- [x] [Program_Licenses.md](https://github.com/StaPH-B/docker-builds/blob/master/Program_Licenses.md) contains the tool(s) used in this PR and has been updated for any missing

### CODE REVIEW

**Comment by kapsakcj** [comment]
I built everything locally, to each of the 3 stages to show the size decrease:
```bash
# Note: 'docker_build' is an alias for 'docker build --progress=plain'
docker_build --target=test -t kapsakcj/samtools:1.18-test samtools/1.18/ && \
docker_build --target=app -t kapsakcj/samtools:1.18-app samtools/1.18/ && \
docker_build --target=builder -t kapsakcj/samtools:1.18-builder samtools/1.18/ && \
\
docker_build --target=test -t kapsakcj/htslib:1.18-test htslib/1.18/ && \
docker_build --target=app -t kapsakcj/htslib:1.18-app htslib/1.18/ && \
docker_build --target=builder -t kapsakcj/htslib:1.18-builder htslib/1.18/ && \
\
docker_build --target=test -t kapsakcj/bcftools:1.18-test bcftools/1.18/ && \
docker_build --target=app -t kapsakcj/bcftools:1.18-app bcftools/1.18/ && \
docker_build --target=builder -t kapsakcj/bcftools:1.18-builder bcftools/1.18/
```
Results:
```bash
$ docker images | grep 'bcftools'
kapsakcj/bcftools    1.18-test       2c282c73decc   35 minutes ago      209MB
kapsakcj/bcftools    1.18-builder    58097f9d6851   35 minutes ago      442MB
kapsakcj/bcftools    1.18-app        9a6cb9b9c648   35 minutes ago      148MB # final uncompressed image size
```
```bash
$ docker images | grep 'samtools'
kapsakcj/samtools   1.18-test        ec985e964df5   3 hours ago         540MB
kapsakcj/samtools   1.18-app         83020475d221   3 hours ago         478MB # final uncompressed image size
kapsakcj/samtools   1.18-builder     2877b4c4ae38   3 hours ago         826MB
```
```bash
$ docker images | grep 'htslib'
kapsakcj/htslib   1.18-test          1c5951dce9d3   2 hours ago      131MB
kapsakcj/htslib   1.18-builder       097e86841ff4   3 hours ago      532MB
kapsakcj/htslib   1.18-app           73429b2aff5a   3 hours ago      131MB # final uncompressed image size
```

**Comment by erinyoung** [comment]
My apologies. It took me a moment to go through GA to find a non-cached test outputs. All three tools appear to work as intended.

From bcftools:
```
#12 [test 2/2] RUN echo "downloading test SC2 BAM and FASTA and running bcftools mpileup and bcftools call test commands..." &&   wget -q https://raw.githubusercontent.com/artic-network/artic-ncov2019/master/primer_schemes/nCoV-2019/V4/SARS-CoV-2.reference.fasta &&   wget -q https://raw.githubusercontent.com/StaPH-B/docker-builds/master/tests/SARS-CoV-2/SRR13957123.primertrim.sorted.bam &&   bcftools mpileup -A -d 200 -B -Q 0 -f SARS-CoV-2.reference.fasta SRR13957123.primertrim.sorted.bam |   bcftools call -mv -Ov -o SRR13957123.vcf
#12 0.061 downloading test SC2 BAM and FASTA and running bcftools mpileup and bcftools call test commands...
#12 0.255 Note: none of --samples-file, --ploidy or --ploidy-file given, assuming all sites are diploid
#12 0.256 [mpileup] 1 samples in 1 input files
#12 0.381 [mpileup] maximum number of reads per input file set to -d 200
#12 0.381 Note: The maximum per-sample depth with -d 200 is 200.0x
#12 DONE 1.2s
```
From samtools
```
...
#12 1.052 # GC-depth. Use `grep ^GCD | cut -f 2-` to extract this part. The columns are: GC%, unique sequence percentiles, 10th, 25th, 50th, 75th and 90th depth percentile
#12 1.052 GCD	0.0	66.667	0.000	0.000	0.000	0.000	0.000
#12 1.052 GCD	38.0	100.000	797.326	797.326	797.326	797.326	797.326
#12 3.055 "plot-SRR13957123.primertrim.sorted-cov.gp" line 116: warning: y2 axis range undefined or overflow
#12 3.271 SRR13957123.consensus.fa
#12 3.271 SRR13957123.consensus.fa.fai
#12 3.271 SRR13957123.primertrim.sorted.bam
#12 3.271 SRR13957123_ampliconstats.txt
#12 3.271 index.html
#12 3.271 nCoV-2019.primer.bed
#12 3.271 plot-SRR13957123.primertrim.sorted-amp.gp
#12 3.271 plot-SRR13957123.primertrim.sorted-amp.png
#12 3.271 plot-SRR13957123.primertrim.sorted-cov.gp
#12 3.271 plot-SRR13957123.primertrim.sorted-cov.png
#12 3.271 plot-SRR13957123.primertrim.sorted-reads.gp
#12 3.271 plot-SRR13957123.primertrim.sorted-reads.png
#12 3.271 plot-SRR13957123.primertrim.sorted-tcoord.gp
#12 3.271 plot-SRR13957123.primertrim.sorted-tcoord.png
#12 3.271 plot-SRR13957123.primertrim.sorted-tdepth.gp
#12 3.271 plot-SRR13957123.primertrim.sorted-tdepth.png
#12 3.271 plot-SRR13957123.primertrim.sorted-tsize.gp
#12 3.271 plot-SRR13957123.primertrim.sorted-tsize.png
#12 3.271 plot-combined-amp.gp
#12 3.271 plot-combined-amp.png
#12 3.271 plot-combined-coverage-1.gp
#12 3.271 plot-combined-coverage-1.png
#12 3.271 plot-combined-depth.gp
#12 3.271 plot-combined-depth.png
#12 3.271 plot-combined-read-perc.gp
#12 3.271 plot-combined-read-perc.png
#12 3.271 plot-combined-reads.gp
#12 3.271 plot-combined-reads.png
#12 3.271 plot-combined-tcoord.gp
#12 3.271 plot-combined-tcoord.png
#12 3.271 plot-combined-tdepth.gp
#12 3.271 plot-combined-tdepth.png
#12 3.271 plot-heat-amp-1.gp
#12 3.271 plot-heat-amp-1.png
#12 3.271 plot-heat-coverage-1-1.gp
#12 3.271 plot-heat-coverage-1-1.png
#12 3.271 plot-heat-read-perc-1.gp
#12 3.271 plot-heat-read-perc-1.png
#12 3.271 plot-heat-read-perc-log-1.gp
#12 3.271 plot-heat-read-perc-log-1.png
#12 3.271 plot-heat-reads-1.gp
#12 3.271 plot-heat-reads-1.png
#12 DONE 3.3s
```

File: samtools/1.18/Dockerfile
Comment (erinyoung) [inline samtools/1.18/Dockerfile:82]: ca-certificates and procps need to be in the app layer. It's required for nextflow (maybe newer versions don't need it, though, so I'm going to be investigation this for a moment).

File: bcftools/1.18/Dockerfile
Comment (erinyoung) [inline bcftools/1.18/Dockerfile:87]: I gave this comment on samtools, but ca-certificates is required for nextflow. I'm going to test these locally to see if it's no longer needed and report back.

**Comment by erinyoung** [comment]
I cannot approve this PR just yet. 

Prior versions of nextflow required procps and ca-certificates when running `-with-tower`. These are often not needed for installing tools and overlooked, which is why I created a GA test to look for them. Since images are built to app and not test, adding procps or ca-certificates in test is likely to appease the GA but may still cause a workflow using `-with-tower` to fail. 

But!

This was several version of nextflow ago. Perhaps that issue has been resolved. I'm going to test these images and report back. If nextflow no longer needs these, then I would like to remove them from GA.

**Comment by erinyoung** [comment]
I can confirm that ca-certificates is no longer required for `-with-tower`. 

I've added an additional change to this PR to include the yaml for the GA workflow so that there is no longer a check for this. 

Since ca-certificates is no longer needed, I've also removed the apt-get stages that include it in the samtools and bcftools dockerfiles.

**Comment by erinyoung** [comment]
Once all checks pass, I'll merge this PR and push to dockerhub and quay

**Comment by erinyoung** [comment]
Looks like wget also needs ca-certificates (there's an option to not need it). I need to make notes of this before I forget all of this again.

I think it's a good idea to remove ca-certificates from the GA test, still.

**Comment by erinyoung** [comment]
It works!!!

I'm going to
- merge this PR
- push the respective samtools, bcftools, and htslib images to dockerhub and quay with the tag 1.18
- overwrite the latest tag for each of these tools

**Review by erinyoung — APPROVED** [review]

**Comment by erinyoung** [comment]
You can check the status of the deployment for these at
- https://github.com/StaPH-B/docker-builds/actions/runs/5955621739
- https://github.com/StaPH-B/docker-builds/actions/runs/5955625295
- https://github.com/StaPH-B/docker-builds/actions/runs/5955623388

**Comment by erinyoung** [comment]
I'm planning on deleting this branch, but I'll keep it for ~10 days in case someone else needs it.

**Comment by kapsakcj** [comment]
thanks for the thorough review, no worries on the delay.

Good to know about nextflow & tower not requiring `wget` & `ca-certificates` anymore.

I'm fine if you want to delete that dev branch, I'm all done with it.

**Pull request merged**
